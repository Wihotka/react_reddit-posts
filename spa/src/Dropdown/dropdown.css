.container {
  width: 100%;
  flex: 1 1 0%;
}

.listContainer {
  position: relative;
  z-index: 3;
}
