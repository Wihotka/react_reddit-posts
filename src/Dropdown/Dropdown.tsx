import React from 'react';
import styles from './dropdown.css';
import { DropdownList } from './DropdownList';

interface IDropdownProps {
  button: React.ReactNode;
  children: React.ReactNode;
  isOpen?: boolean;
  onOpen?: () => void;
  onClose?: () => void;
}

const NOOP = () => {};

export function Dropdown({
  button,
  children,
  isOpen,
  onOpen = NOOP,
  onClose = NOOP
}: IDropdownProps) {
  const [isDropdownOpen, setIsDropdownOpen] = React.useState(isOpen);

  React.useEffect(() => setIsDropdownOpen(isOpen), [isOpen]);
  React.useEffect(() => isDropdownOpen ? onOpen() : onClose(), [isDropdownOpen]);

  const handleOpen = () => {
    if (isOpen === undefined) {
      setIsDropdownOpen(!isDropdownOpen);
    }
  }

  return (
    <div className={styles.container}>
      <div onClick={handleOpen}>
        {button}
      </div>
      <div className={styles.listContainer} id="dropdown-root">
        {isDropdownOpen &&
          <DropdownList children={children} onElemClick={() => setIsDropdownOpen(false)} />
        }
      </div>
    </div>
  );
}
