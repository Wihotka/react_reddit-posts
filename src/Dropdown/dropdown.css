.container {
    width: 100%;
    flex: 1 1 0%;
}

.listContainer {
    position: relative;
    z-index: 3;
}

.list {
    position: absolute;
    top: 4px;
    left: 0;
    width: 100%;
    height: 100%;
    opacity: 0;
    visibility: hidden;
    transition: 150ms ease-in-out;
}

.listActive {
    opacity: 1;
    visibility: visible;
}
