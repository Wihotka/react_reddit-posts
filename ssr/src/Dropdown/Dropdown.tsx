import React, {ReactNode} from 'react';
import styles from './dropdown.css';
import ReactDOM from "react-dom";

interface IDropdownProps {
    button: React.ReactNode;
    children: React.ReactNode;
    portalNode?: HTMLDivElement;
    isOpen?: boolean;
    onOpen?: () => void;
    onClose?: () => void;
}

const NOOP = () => {};

export function Dropdown({button, children, isOpen, onOpen = NOOP, onClose = NOOP}: IDropdownProps) {
    const [isDropdownOpen, setIsDropdownOpen] = React.useState(isOpen);
    React.useEffect(() => setIsDropdownOpen(isOpen), [isOpen]);
    React.useEffect(() => isDropdownOpen ? onOpen() : onClose(), [isDropdownOpen]);

    const handleOpen = () => {
        if (isOpen === undefined) {
            setIsDropdownOpen(!isDropdownOpen);
        }
    }

    const handleClose = () => {
        setIsDropdownOpen(false);
    }

    const node = document.getElementById('react-dropdown');

    if(!node) return null;

    return (
        <div className={styles.container}>
            <div onClick={handleOpen}>
                {button}
            </div>
            {isDropdownOpen && ReactDOM.createPortal((
                <div onClick={handleClose} className={styles.listContainer}>
                    {children}
                </div>
            ), node)}
        </div>
    );
}
