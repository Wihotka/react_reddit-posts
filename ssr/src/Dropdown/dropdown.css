.container{
    width: 100%;
    flex: 1 1 0%;
}

.listContainer{
    position: fixed;
    top: 15%;
    right: 15%;
    z-index: 99999;
}

.list {
    position: absolute;
    overflow: hidden;
    left: 0;
    top: 0;
}
