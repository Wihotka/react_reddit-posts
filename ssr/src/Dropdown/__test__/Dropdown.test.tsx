import React from 'react';
import {shallow} from "enzyme";
import {Dropdown} from "../Dropdown";
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

Enzyme.configure({ adapter: new Adapter() });

describe('Dropdown', () => {
    test('should render', () => {
        const wrapper = shallow(<Dropdown button={<div/>} children={<button/>}/>);
        expect(wrapper).toBeDefined();
        expect(wrapper.find('div.container').isEmptyRender()).toBeFalsy();
    })

    test('shold render (snapshot)', () => {
        const wrapper = shallow(<Dropdown button={<div/>} children={<button/>}/>);

        expect(wrapper).toMatchSnapshot();
    })
});
