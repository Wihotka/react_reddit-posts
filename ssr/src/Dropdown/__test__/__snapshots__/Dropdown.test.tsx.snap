// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown shold render (snapshot) 1`] = `
<div
  className="container"
>
  <div
    onClick={[Function]}
  >
    <div />
  </div>
</div>
`;
